### DISCUSSION TITLE
How to create content with custom routes?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#9260] [Q&A]
**Goal**
I am trying to [recreate this page ](https://realestatebyareen.com/homes-for-sale-in-las-vegas-16/1) using OC. I am a bit lost and not sure what is the correct approach for this.

I split that page into multiple contact part/fields as explained in the screenshot below
![image](https://user-images.githubusercontent.com/24724371/115773968-d5449c80-a365-11eb-8405-1edc47458a8c.png)

Note the middle form need to be editable in the admin page. One may want to add more/remove filters from the page.

Also, the route needs to fit this pattern `/homes-for-sale-in-{CITY_NAME_SLUG}-{CITY_ID}/{PAGE_NUMBER}`

**Initial Approach**
I am thinking to create a new content type called `HomeForSaleByCity` and each city will eventually be a content item under the `HomeForSaleByCity` type. The `HomeForSaleByCity` type will be migrated with  `TitlePart`, `SitePagePart`, and `WidgetListPart`.

I am also thinking that I should be able to generate the route by using a custom controller/action. like this

```
public class HomeForSaleByCityController : Controller
{
   [Route("/homes-for-sale-in-{city:slugify}-{id:int}/{page:int?}", Name = "HomesForSaleByCity")]
    public IActionResult Display(string city, int id, int? page)
    {
        return View();
    }
}
```

However, I am puzzled on how to render that filter you see in the middle of the screenshot above. For each filter on that form, I created an attachable content part but not sure how to put it all together.

**Question**
Is my initial approach correct? Is there a better approach to accomplish this?
How can I add the filters to the page?

**Comment by sebastienros** [comment]
I think the best approach is to use the a lucene index and build a lucene query from the querystring parameters.
If you prefer to do everything with SQL then you might need to create a custom Index Provider that will store the properties of your content items in dedicated db fields so you can build a custom query.

Your content type should be Home, not "bycity" or "forsale", these are just properties of your Home content items (fields). Then you can query all the content items "for sale", or "by city" or even with radius since the GeoPoint field is available now. A custom controller will still be required.

**Reply by MikeAlhayek** [reply]
@sebastienros Thanks for your help! The home info is actually will be fetched using an API and not hosted locally. The question is really about how to render all that info you see on the screen. What I want to store locally is info about the page like 
 1. Title & Subtitle
 2. Page Content
 3. Page Description
 3. What search filters to show on that page.

Each of these pages will represent a different city and the URL will change to reflect the name and the city id in the url as per the pattern above.

Similarly, I have other search pages like [/search-by-county/nevada/nye](https://realestatebyareen.com/search-by-county/nevada/nye) similar to the above, this will have a way to add/edit pages for this type of search. There are other types of search here too and each has a unique URL pattern like search by zip code, search near by, search by school.

I am guessing the only way to accommodate this is created a different controller/action for each search/route. Is that true assumption? How would I render the search form (where user can change the filters on that form) on the page using custom controller per search?

**Reply by MikeAlhayek** [reply]
@deanmarcussen @jtkech any thoughts on how to approach this problem?

**Comment by deanmarcussen** [comment]
I agree with @sebastienros 

It's one controller, a Home controlller, which takes a model from the query string / route values of filters, and from those filters makes whatever kind of query you require from the external api to populate a list of results.

From reading your notes above, it sounds like you are struggling to figure out how to produce a controller with filters?

Perhaps a helpful example might be either the users admin controller or the contents admin list controller, which used different filter providers to generate its filter ui.

Here is the pr where I added user filters (they're probably simpler to look at than the contents list)
https://github.com/OrchardCMS/OrchardCore/pull/8800

There'll be a demo on the lombiq youtube channel I imagine.

**Reply by MikeAlhayek** [reply]
@deanmarcussen I am struggling to understand how to renders the search-form on the screen along with unique content per page. When the users visits the page, I'll need to render the search-form along with page info (details, title...). 

If I understand the recommended approach correctly, I would want to create the following controller and model, correct?

```
public class SearchController : Controller
{
    [Route("/homes-for-sale-in-{city:slugify}-{cityId:int}/{page:int?}", Name = "HomesForSaleByCity")]
    [Route("/search-by-couty-{state:slugify}-{county:slugify}/{page:int?}", Name = "HomesForSaleByCounty")]
    public IActionResult Display(HomesQueryViewModel viewModel, PagerParameters pagerParameters)
    {
    	// somehow generate the content
        return View();
    }
}

public class HomesQueryViewModel
{
	public string City { get; set; }
	public string CityId { get; set; }
	public string State { get; set; }
	public string County { get; set; }

	// any other possible filter like zip code, school, etc...
}
```

In the mentioned PR, the `UserIndexOptions` is predefined view-model instead of a dynamic form. But in my case, the form, is something I want the admin to be able to define from the admin panel. If a admin wants to add new drop down menu as a filter, I want to them to be able to do that. 

Really I am struggling with two things here. 

 1. How can I add different content per page? The page content is captured by using the `SitePagePart`. For example, the URL `/home-for-sale-in-los-angeles-1` will have different page description and details than `/home-for-sale-las-vegas-10`. The search-form will look the same in this case because the route is shared. However, the `/search-by-county/nevada/nye` will most likely have a completely different search form since its coming from a different route. Again, these search-form will need to be controlled by the admin page so the admin can define add/remove filters as he/she sees fit.
 2.  How can I show the custom search-form "content-type" to the result page.

**Reply by deanmarcussen** [reply]
Find the video of the demo. Have a read of the code. They are all extensible. See the example of the contents admin list and the optional taxonomy filters.

If it’s all driven by an external api then much will depend on how the query to the external api needs to be driven.
